File: Ruby/Chapter01/knock04.rb
# Decompose this sentence that is  "Hi He Lied Because Boron Could Not Oxidize Fluorine. New Nations Might Also Sign Peace Security Clause. Arthur King Can.", 
# and get 1 or 2 characters from head.
# In addition, associative array.: from characters to order.
#



#init
sentence = "Hi He Lied Because Boron Could Not Oxidize Fluorine. New Nations Might Also Sign Peace Security Clause. Arthur King Can."

num = 0
num_gene = Enumerator.new do |y|
  [1, 5, 6, 7, 8, 9, 15, 16, 19, 0].each{|x| y.yield(x)}
end
next_num = num_gene.next

symbol = Hash.new

# process
sentence.split(' ').each{|word|
  num += 1
  get_num = 2
  if num == next_num then
    next_num = num_gene.next
    get_num -= 1
  end
  #puts("order: #{num}, symbol: #{word.slice(0, get_num)}")
  symbol[word.slice(0, get_num)] = num
}

symbol.each_pair {|sym, order| 
  puts("sym : #{sym}, num : #{order}")
}

File: Csharp/chapter01/knock04.cs
/*
 * Chapter01, knock04
 * Task: make map
 */

using System;
using System.Collections.Generic;

public class ElementMap {
    static public void Main() {
        string sentence = "Hi He Lied Because Boron Could Not Oxidize Fluorine. New Nations Might Also Sign Peace Security Clause. Arthur King Can.";
        PrintMap(MakeMap(sentence));
    }


    public static Dictionary<string, int> MakeMap(string str) {
        string[] delimiter= new string[] {" "};
        string[] words = str.Split(delimiter, StringSplitOptions.None);
        Dictionary<string, int> elementMap = new Dictionary<string, int>();
        int[] nums = new int[]{1, 5, 6, 7, 8, 9, 15, 16, 19, -1};
        int j = 0, end = 0;
        for (int i=0; i < words.Length; i++) {
            if (nums[j] == i + 1) {
                end = 1;
                j++;
            } else {
                end = 2;
            }
            elementMap[words[i].Substring(0, end)] = i + 1;
        }
        return elementMap;
    }

    public static void PrintMap(Dictionary<string, int> map) {
        foreach (KeyValuePair<string, int> pair in map) {
            Console.WriteLine(string.Format("{0} \t {1}", pair.Key, pair.Value));
        }
    }

}

